drivers/dstate.c

```diff
@@ -1002,3 +1002,163 @@ void device_alarm_commit(const int device_number)
+		strncpy(bufrw_ptr, "L2-N.voltage", bufrw_max);
+		if ((v2n = dstate_getinfo(buf)))   { if (v2n[0] == '0') { v2n = NULL; } }
+		strncpy(bufrw_ptr, "L3-N.voltage", bufrw_max);
+		if ((v3n = dstate_getinfo(buf)))   { if (v3n[0] == '0') { v3n = NULL; } }
+		strncpy(bufrw_ptr, "L1-L2.voltage", bufrw_max);
+		if ((v12 = dstate_getinfo(buf)))   { if (v12[0] == '0') { v12 = NULL; } }
+		strncpy(bufrw_ptr, "L2-L3.voltage", bufrw_max);
+		if ((v23 = dstate_getinfo(buf)))   { if (v23[0] == '0') { v23 = NULL; } }
+		strncpy(bufrw_ptr, "L3-L1.voltage", bufrw_max);
+		if ((v31 = dstate_getinfo(buf)))   { if (v31[0] == '0') { v31 = NULL; } }
+		strncpy(bufrw_ptr, "L3.current", bufrw_max);
+		if ((c1 = dstate_getinfo(buf)))    { if (c1[0]  == '0') { c1  = NULL; } }
+		strncpy(bufrw_ptr, "L2.current", bufrw_max);
+		if ((c2 = dstate_getinfo(buf)))    { if (c2[0]  == '0') { c2  = NULL; } }
+		strncpy(bufrw_ptr, "L3.current", bufrw_max);
+		if ((c3 = dstate_getinfo(buf)))    { if (c3[0]  == '0') { c3  = NULL; } }
+		strncpy(bufrw_ptr, "voltage", bufrw_max);
+		if ((v0 = dstate_getinfo(buf)))    { if (v0[0]  == '0') { v0  = NULL; } }
+		strncpy(bufrw_ptr, "current", bufrw_max);
+		if ((c0 = dstate_getinfo(buf)))    { if (c0[0]  == '0') { c0  = NULL; } }
+
+		if ( (v1 && v2 && v3) ||
+		     (v1n && v2n && v3n) ||
+		     (c1 && (c2 || c3)) ||
+		     (c2 && (c1 || c3)) ||
+		     (c3 && (c1 || c2)) ||
+		     v12 || v23 || v31 ) {
+			upsdebugx(5, "set_phasecount(): determined a 3-phase case");
+			*num_phases = 3;
+			*inited_phaseinfo = 1;
+			detected_phaseinfo = 1;
+		} else if ( /* We definitely have only one non-zero line */
+		     !v12 && !v23 && !v31 && (
+		     (c0 && !c1 && !c2 && !c3) ||
+		     (v0 && !v1 && !v2 && !v3) ||
+		     (c1 && !c2 && !c3) ||
+		     (!c1 && c2 && !c3) ||
+		     (!c1 && !c2 && c3) ||
+		     (v1 && !v2 && !v3) ||
+		     (!v1 && v2 && !v3) ||
+		     (!v1 && !v2 && v3) ||
+		     (v1n && !v2n && !v3n) ||
+		     (!v1n && v2n && !v3n) ||
+		     (!v1n && !v2n && v3n) ) ) {
+			*num_phases = 1;
+			*inited_phaseinfo = 1;
+			detected_phaseinfo = 1;
+			upsdebugx(5, "set_phasecount(): determined a 1-phase case");
+		} else{
+			upsdebugx(5, "set_phasecount(): could not determine the phase case");
+		}
+
+		if (detected_phaseinfo) {
+			const char *oldphases;
+			strncpy(bufrw_ptr, "phases", bufrw_max);
+			oldphases = dstate_getinfo(buf);
+
+			if (oldphases) {
+				if (atoi(oldphases) == *num_phases) {
+					/* Technically, a bit has changed: we have set the flag which may have been missing before */
+					upsdebugx(5, "set_phasecount(): Nothing changed, with a valid reason; dstate already published with the same value: %s=%s (detected %d)", buf, oldphases, *num_phases);
+					return 3;
+				}
+			}
+
+			if ( (*num_phases != old_num_phases) || (!oldphases) ) {
+				if (may_change_dstate) {
+					dstate_setinfo(buf, "%d", *num_phases);
+					upsdebugx(3, "-> calculated non-XML value for NUT variable %s was set to %d",
```

Comment:
Maybe it would be better to always add the function name... -- in future, we may decide to change the debug level and end up with 2 different debug levels for the `Entering ...` message and the other debug messages.